##### 090200354, Beyza Kopuz

---
title: "Task 2"
subtitle: "Data Pre-processing"
author: "Beyza Kopuz"
date: "April 3, 2023"
---

**Task 2:** Find a data set which is suitable for regression analysis and consists of a mix of numerical, nominal, and ordinal variables. Look for the cases, where at least one of the variable has missing values. If not, you can randomly delete a very small portion of one of the variables.
Design a machine learning pipeline where you scale the numerical features and encode the nominal and ordinal features along with imputing the missing values.

#### Instructions to follow:

- Include all your codes here. Be sure that your code is CLEAN, READABLE, and REPRODUCIBLE.
- Put your data set into a **datasets** folder.
- Put your images (if available) into an **images** folder.
- Please return a NICE and CLEAR homework. Otherwise, it will not be graded.
- Please write YOUR OWN code. **DO NOT copy** my codes or someone else's codes.

## Data Decription

- Import and introduce your data set here.

In [1]:
import pandas as pd
import numpy as np

In [2]:
house = pd.read_csv(r'C:\Users\beyza\OneDrive\Masaüstü\datasets\Housing.csv')
house

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420         4          2        3      yes        no       no   
1    12250000  8960         4          4        4      yes        no       no   
2    12250000  9960         3          2        2      yes        no      yes   
3    12215000  7500         4          2        2      yes        no      yes   
4    11410000  7420         4          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000         2          1        1      yes        no      yes   
541   1767150  2400         3          1        1       no        no       no   
542   1750000  3620         2          1        1      yes        no       no   
543   1750000  2910         3          1        1       no        no       no   
544   1750000  3850         3          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

This 'Housing' data set can be found in the 'datasets' folder and at the end of this notebook on References list. 
It has 13 columns and 545 rows.

The column names are as follows:
- price: Price of the house. (This is the target value.)
- area: Area of the house.
- bedrooms: Number of bedrooms the house has.
- bathrooms: Number of bathrooms the house has.
- stories: Number of the stories in the house.
- mainroad: Whether the house is connected to the main road.
- guestroom: Whether the house has a guestroom.
- basement: Whether the house has a basement.
- hotwaterheating: Whether the house has hot water heater.
- airconditioning: Whether the house has air conditioning.
- parking: Number of house parking.
- prefarea: Whether the house is in a preferred area. (This is my interpretation. Source doesn't explain this further.)
- furnishingstatus: Furnishing status of the house.

In [3]:
house.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


This data set has no not-available entries. In order to do the asked task, I will have to delete a small portion of one of the variables. I choose to delete from the 'bedrooms' variable.

In [4]:
bedr = house['bedrooms']
df = pd.DataFrame(bedr, columns=['bedrooms'])

In [5]:
def delete(col):
    col.loc[col.sample(frac=0.045).index] = np.nan
    return col

df.apply(delete, axis=0)

bedrooms
0         4.0
1         4.0
2         3.0
3         4.0
4         4.0
..        ...
540       2.0
541       3.0
542       2.0
543       3.0
544       3.0

[545 rows x 1 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   bedrooms  520 non-null    float64
dtypes: float64(1)
memory usage: 4.4 KB


In [7]:
house['bedrooms'] = df['bedrooms']
house

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0    13300000  7420       4.0          2        3      yes        no       no   
1    12250000  8960       4.0          4        4      yes        no       no   
2    12250000  9960       3.0          2        2      yes        no      yes   
3    12215000  7500       4.0          2        2      yes        no      yes   
4    11410000  7420       4.0          1        2      yes       yes      yes   
..        ...   ...       ...        ...      ...      ...       ...      ...   
540   1820000  3000       2.0          1        1      yes        no      yes   
541   1767150  2400       3.0          1        1       no        no       no   
542   1750000  3620       2.0          1        1      yes        no       no   
543   1750000  2910       3.0          1        1       no        no       no   
544   1750000  3850       3.0          1        2      yes        no       no   

    hotwaterheating airconditioning  parking prefarea furnishingstatus  
0                no             yes        2      yes        furnished  
1                no             yes        3       no        furnished  
2                no              no        2      yes   semi-furnished  
3                no             yes        3      yes        furnished  
4                no             yes        2       no        furnished  
..              ...             ...      ...      ...              ...  
540              no              no        2       no      unfurnished  
541              no              no        0       no   semi-furnished  
542              no              no        0       no      unfurnished  
543              no              no        0       no        furnished  
544              no              no        0       no      unfurnished  

[545 rows x 13 columns]

In [8]:
house.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price             545 non-null    int64  
 1   area              545 non-null    int64  
 2   bedrooms          520 non-null    float64
 3   bathrooms         545 non-null    int64  
 4   stories           545 non-null    int64  
 5   mainroad          545 non-null    object 
 6   guestroom         545 non-null    object 
 7   basement          545 non-null    object 
 8   hotwaterheating   545 non-null    object 
 9   airconditioning   545 non-null    object 
 10  parking           545 non-null    int64  
 11  prefarea          545 non-null    object 
 12  furnishingstatus  545 non-null    object 
dtypes: float64(1), int64(5), object(7)
memory usage: 55.5+ KB


This is the new (removed) data set that I am going to use. 

### Data Splitting

Performing data splitting at the beginning of the work flow is very important as it prevents data leakage.

In [9]:
#Separating the target value and the feature matrix.
house_X = house.drop(columns=['price'], axis=1)
house_y = house['price']

In [10]:
#Data splitting
from sklearn.model_selection import train_test_split

house_X_train, house_X_test, house_y_train, house_y_test = train_test_split(house_X, house_y, test_size=0.2, random_state=1989)

## Pipeline Design

- Design your machine learning pipeline here.

In [11]:
#sklearn returns numpy arrays. but we need pandas dataframes. with this line of code, everything will be returned in pandas dataframe. 
from sklearn import set_config
set_config(transform_output="pandas") 

I am going to design 3 different pipelines each for numerical, nominal and ordinal features. To do that, I need to separate the features in the house_X_train to numerical, nominal and ordinal. As I've learned in further steps, the ColumnTransformer doesn't accept column names with string. So I'm listing the columns by indices here.

In [12]:
numerical_features = house_X_train.columns[[0,1,2,3,9]]
nominal_features = house_X_train.columns[[4,5,6,7,8,10]]
ordinal_features = house_X_train.columns[[11]]

In [13]:
numerical_features

Index(['area', 'bedrooms', 'bathrooms', 'stories', 'parking'], dtype='object')

In [14]:
nominal_features

Index(['mainroad', 'guestroom', 'basement', 'hotwaterheating',
       'airconditioning', 'prefarea'],
      dtype='object')

In [15]:
ordinal_features

Index(['furnishingstatus'], dtype='object')

- Instantiate the imputers.

In [16]:
from sklearn.impute import SimpleImputer

numerical_imputer = SimpleImputer(strategy="mean")
nominal_imputer = SimpleImputer(strategy="most_frequent")
ordinal_imputer = SimpleImputer(strategy="most_frequent")

- Instantiate the scaler and encoders.

In [17]:
furnishing_cat = ["unfurnished","semi-furnished","furnished"] #Orders are 0, 1, 2 respectively.

In [18]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder

numerical_preprocessor = StandardScaler()
nominal_preprocessor = OneHotEncoder(drop='if_binary', sparse_output=False)  
ordinal_preprocessor = OrdinalEncoder(categories=[furnishing_cat])

- Create preprocessing pipelines for numerical, nominal and ordinal data separately.

In [19]:
from sklearn.pipeline import make_pipeline

numerical_pipeline = make_pipeline(
    numerical_imputer, 
    numerical_preprocessor, 
)

nominal_pipeline = make_pipeline(
    nominal_imputer,
    nominal_preprocessor
)

ordinal_pipeline = make_pipeline(
    ordinal_imputer,
    ordinal_preprocessor
)

- Combining the pipelines together.

In [20]:
from sklearn.compose import ColumnTransformer

#full transformer
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numerical_pipeline, numerical_features),
        ("nom", nominal_pipeline, nominal_features),
        ("ord", ordinal_pipeline, ordinal_features)
    ]
)

- Creating the final pipeline.

In [21]:
from sklearn.linear_model import LinearRegression

pipe = make_pipeline(
    preprocessor, 
    LinearRegression()
)

In [22]:
%%capture 
#Output of this line doesn't look pretty so I'm hiding the output for the clean look.
pipe

## Pipeline Results

- Put your regression analysis (which is based on a pipeline object) results here. Comment on the results.

- Applying the pipeline to the training data **only**.

In [23]:
%%capture 
#Output of this line doesn't look pretty so I'm hiding the output for the clean look.
pipe.fit(house_X_train, house_y_train)

- Predicting the test values.

In [24]:
%%capture 
#Output of this line doesn't look pretty so I'm hiding the output for the clean look.
pipe.predict(house_X_test)

- Finally, checking R^2 score to see how good this model is.

In [25]:
print('R^2 score of the test data: %.2f' % pipe.score(house_X_test, house_y_test)) 

R^2 score of the test data: 0.72


I think the R^2 score is close enough to 1 and this linear model is a good model for this data set.

## References
- https://www.kaggle.com/datasets/yasserh/housing-prices-dataset
- https://scikit-learn.org/stable/auto_examples/compose/plot_column_transformer_mixed_types.html
- https://inria.github.io/scikit-learn-mooc/python_scripts/03_categorical_pipeline_column_transformer.html
- https://stackoverflow.com/questions/61641852/what-is-the-valid-specification-of-the-columns-needed-for-sklearn-classifier-p
- https://github.com/gulinan Gül İnan Lecture Slides. 